handlers/rsa.go

```diff
@@ -0,0 +1,211 @@
+func (rh *RsaHandler) GetName() string {
+	return "rsa"
+}
+
+// Init initializes the handler.
+// Takes one argument for persistent data directory.
+func (rh *RsaHandler) Init(opts ...string) (err error) {
+	rh.kv, err = metadb.New(db.TypePebble, filepath.Clean(opts[0]))
+	return err
+}
+
+// Auth is the handler for the rsa handler
+func (rh *RsaHandler) Auth(r *http.Request,
+	ca *csp.Message, pid []byte, st string) (bool, string) {
+	log.Infof(r.UserAgent())
+
+	voterId, message, signature, err := parseRsaAuthData(ca.AuthData)
+	if err != nil {
+		log.Error(err)
+		return false, err.Error()
+	}
+
+	rsaPublickey, err := parseRsaPublicKey()
+	if err != nil {
+		log.Error(err)
+		return false, err.Error()
+	}
+
+	// Verify signature
+	if err := validateRsaSignature(signature, message, rsaPublickey); err != nil {
+		log.Warnf("invalid signature: %s\n", err)
+		return false, "invalid signature"
+	}
+
+	if st == csp.SignatureTypeSharedKey {
+		return true, "please, do not share the key"
+	}
+	if rh.exist(voterId) {
+		log.Warnf("%s already registered", voterId)
+		return false, "already registered"
+	}
+
+	rh.addKey(voterId, nil)
+	log.Infof("new user registered with id %s", voterId)
+
+	return true, ""
+}
+
+// RequireCertificate must return true if the auth handler requires some kind of client
+// TLS certificate. If true then CertificateCheck() and HardcodedCertificate() methods
+// must be correctly implemented. Else both function can just return true and nil.
+func (rh *RsaHandler) RequireCertificate() bool {
+	return false
+}
+
+// CertificateCheck is used by the Auth handler to ensure a specific certificate is
+// added to the CA cert pool on the HTTP/TLS layer (optional).
+func (rh *RsaHandler) CertificateCheck(subject []byte) bool {
+	return true
+}
+
+// Certificates returns a hardcoded CA certificated that will be added to the
+// CA cert pool by the handler (optional).
+func (rh *RsaHandler) Certificates() [][]byte {
+	return nil
+}
+
+// Internal data handlers
+
+func parseRsaPublicKey() (*rsa.PublicKey, error) {
+	if len(rsaPubKeys) < 1 {
+		return nil, errors.Errorf("no public keys")
+	}
+	// Using the first one available so far
+	block, _ := pem.Decode([]byte(rsaPubKeys[0]))
+	if block == nil {
+		log.Warnf("failed to parse the public key")
+		return nil, errors.Errorf("failed to parse the public key")
+	}
+
+	parsedKey, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	switch parsedKey.(type) {
+	case *rsa.PublicKey:
+		break
+	case *dsa.PublicKey:
+	case *ecdsa.PublicKey:
+	case ed25519.PublicKey:
+	default:
+		log.Warnf("cannot parse the public key")
+		return nil, errors.Errorf("cannot parse the public key")
+	}
+	return parsedKey.(*rsa.PublicKey), nil
+}
+
+// parseRsaAuthData transforms the incoming authData string array and returns a digested output
+// of the relevant parameters for the handler
+func parseRsaAuthData(authData []string) ([]byte, []byte, []byte, error) {
```

Comment:
Maybe these three return []byte values could be packed inside a struct, that would be more idiomatic and better for understanding the code

Reply:
Done